rslib/src/browser_rows.rs

```diff
@@ -0,0 +1,356 @@
+// Copyright: Ankitects Pty Ltd and contributors
+// License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
+
+use std::sync::Arc;
+
+use itertools::Itertools;
+
+use crate::err::{AnkiError, Result};
+use crate::i18n::{tr_args, I18n, TR};
+use crate::{
+    card::{Card, CardID, CardQueue, CardType},
+    collection::Collection,
+    decks::{Deck, DeckID},
+    notes::Note,
+    notetype::{CardTemplate, NoteType, NoteTypeKind},
+    scheduler::{timespan::time_span, timing::SchedTimingToday},
+    template::RenderedNode,
+    text::{extract_av_tags, html_to_text_line},
+    timestamp::{TimestampMillis, TimestampSecs},
+};
+
+const CARD_RENDER_COLUMNS: [&str; 2] = ["question", "answer"];
+
+#[derive(Debug, PartialEq)]
+pub struct Row {
+    pub cells: Vec<Cell>,
+    pub color: Color,
+    pub font: Font,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Cell {
+    pub text: String,
+    pub is_rtl: bool,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum Color {
+    Default,
+    Marked,
+    Suspended,
+    FlagRed,
+    FlagOrange,
+    FlagGreen,
+    FlagBlue,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Font {
+    pub name: String,
+    pub size: u32,
+}
+
+struct RowContext<'a> {
+    col: &'a Collection,
+    card: Card,
+    note: Note,
+    notetype: Arc<NoteType>,
+    deck: Option<Deck>,
+    original_deck: Option<Option<Deck>>,
+    i18n: &'a I18n,
+    timing: SchedTimingToday,
+    render_context: Option<RenderContext>,
+}
+
+/// The answer string needs the question string but not the other way around, so only build the
+/// answer string when needed.
+struct RenderContext {
+    question: String,
+    answer_nodes: Vec<RenderedNode>,
+}
+
+fn card_render_required(columns: &[String]) -> bool {
+    columns
+        .iter()
+        .any(|c| CARD_RENDER_COLUMNS.contains(&c.as_str()))
+}
+
+impl Collection {
+    pub fn browser_row_for_card(&mut self, id: CardID) -> Result<Row> {
+        let columns: Vec<String> = self.get_config_optional("activeCols").unwrap_or_else(|| {
+            vec![
```

Comment:
It's best to define defaults in a file in the config/ folder, so that the rest of the codebase only has to call a function to ensure it'll get a valid results with a valid default. For bools/strings we have dedicated enums; other types of content currently need their own getter/setter funcs

Reply:
+ the config key should be listed in the enum so we don't need to identify it by a string - will fix this up.

Reply:
Right, I noticed that we currently define the defaults independently in the frontend and backend which will be solved when the frontend can access a proper getter.
Initially I thought, we would make the columns deserialisable later anyway.

Reply:
Currently only the strings and bools are exposed to the frontend - maybe we could have a different enum for arbitrary json values.

It bugs me that we're allocating strings for every row, even though it's probably lost in the noise floor of Qt's rendering :-) We could potentially get serde to deserialize a list of int|str into an enum, and just preserve unrecognized strings for the benefit of add-ons, though we might want to use a new config key if doing that so downgrading doesn't cause issues.

Reply:
If serialise into and deserialise from the old string list, why would that be a problem to downgrading? I thought the benefit of this approach would be that it's fully backwards compatible.
However, I assumed we would have to write a lot of custom serde code for that and that updating the databse when up- or downgrading would be easier.
Now then you mention a separate key: Can we maybe use a different key with enum columns from now on, but if it's absent, look for the old key and transform the strs into enum variants? As far as I can see, neither downgrading steps nor custom deserialisation would be necessary.

Reply:
The advantage of serializing into a list of numbers (for known keys) is that it's cheaper to read back later, but we'd still need to serialize Column::Other(String) into a string value. Serde can handle a mix of strings and integers with "untagged", but I don't think it could handle a value like Other(String) - we'd probably need another enum, or need to parse into a number|str first, and convert that number to the enum variant. I'm starting to gravitate back towards my initial impression of "let's just leave it for now"  ^_^;

Reply:
Gotcha. One last suggestion before I shut up about it ;) :
Maintain two config entries, one for the backend, a list of numbers deserialisable into an enum where all unrecognised names are mapped to the same number, and one for the frontend, a list of strings. We only a need a special setter that saves both entries from a list of strings.

Reply:
Add-ons and previous versions could update one without changing the other though, meaning the backend would need to do extra validation to make sure everything matched

Reply:
Yeah, the config setter on front- or backend would have to check for the "activeCols" key and call the special setter to avoid that. Alternatively, the browser model could set the backend entry whenever it gets the frontend entry.